### ISSUE TITLE
Failed to adjust DRBD resource on toggle-disk primary operation

### ISSUE THREAD

**Issue by kvaps** [LINBIT/linstor-server#84]
Hi, today I was faced with the next situation:

```
# linstor v l | grep opennebula-data-opennebula-db
| pve1   | opennebula-data-opennebula-db-0                | pve                  | 0        | 1169    | /dev/drbd1169 | 10.00 GiB  | Unused |          UpToDate |
| pve2   | opennebula-data-opennebula-db-0                | pve                  | 0        | 1169    | /dev/drbd1169 | 10.00 GiB  | InUse  |          UpToDate |
| pve3   | opennebula-data-opennebula-db-0                | pve                  | 0        | 1169    | /dev/drbd1169 | 10.00 GiB  | Unused |          UpToDate |
```

```
# linstor r toggle-disk pve2 opennebula-data-opennebula-db-0 -d
SUCCESS:
    Removal of disk from resource 'opennebula-data-opennebula-db-0' on node 'pve2' registered
SUCCESS:
    Removed disk on 'pve2'
SUCCESS:
    Notified 'pve3' that disk has been removed on 'pve2'
SUCCESS:
    Notified 'pve1' that disk has been removed on 'pve2'
root@linstor-controller-0:/# linstor r toggle-disk pve2 opennebula-data-opennebula-db-0 -s data
SUCCESS:
    Addition of disk to resource 'opennebula-data-opennebula-db-0' on node 'pve2' registered
SUCCESS:
    Prepared 'pve3' to expect disk on 'pve2'
SUCCESS:
    Prepared 'pve1' to expect disk on 'pve2'
ERROR:
    (Node: 'pve2') Failed to adjust DRBD resource opennebula-data-opennebula-db-0
Show reports:
    linstor error-reports show 5DA5C3B1-F21CA-000002
ERROR:
    (Node: 'pve1') Failed to adjust DRBD resource opennebula-data-opennebula-db-0
Show reports:
    linstor error-reports show 5DAF8DF8-8A94D-000010
```

```
# linstor v l | grep opennebula-data-opennebula-db
| pve1   | opennebula-data-opennebula-db-0                | pve                  | 0        | 1169    | /dev/drbd1169 | 10.00 GiB  | Unused |           UpToDate |
| pve2   | opennebula-data-opennebula-db-0                | data                 | 0        | 1169    | /dev/drbd1169 |            | InUse  |           Diskless |
| pve3   | opennebula-data-opennebula-db-0                | pve                  | 0        | 1169    | /dev/drbd1169 | 10.00 GiB  | Unused |           Outdated |
```

[ErrorReport-5DA5C3B1-F21CA-000002.log](https://github.com/LINBIT/linstor-server/files/3766689/ErrorReport-5DA5C3B1-F21CA-000002.log)

Why it becomes to use the `data` pool, it should use `DfltDisklessStorPool` instad, isn't it?

**this information may be useful:**

Before that I was migrating another resource from `pve` to `diskless` pool and then from `diskless` to `data` using `toggle-disk` command

**Comment by raltnoeder** [comment]
You issued the commands:
```
linstor r toggle-disk pve2 opennebula-data-opennebula-db-0 -d
linstor r toggle-disk pve2 opennebula-data-opennebula-db-0 -s data
```
The first command removed the disk of the resource `opennebula-data-opennebula-db-0` on node `pve2` and selected the default diskless storage pool instead.
The second command was supposed to create and attach a new backing disk for the resource `opennebula-data-opennebula-db-0` on node `pve2` in the storage pool `data`, which was manually selected by specifying the `-s data` option.

The more interesting question is why it failed, and why drbdadm apparently decided to run a couple `drbdsetup disconnect` commands.

**Comment by kvaps** [comment]
> You issued the commands:
> 
> ```
> linstor r toggle-disk pve2 opennebula-data-opennebula-db-0 -d
> linstor r toggle-disk pve2 opennebula-data-opennebula-db-0 -s data
> ```

That's true, sorry I didn't wake up today 🙂

**Comment by kvaps** [comment]
**UPD:** after that situation my controller is totally stuck

Eg. I can't run any commands like `linstor c lp`, linstor r l` because they are hangs forever.

Log from linstor-controller:
```
Version:            1.1.2 (6fa27183eed583088e07b757cdef61ba13a14854)
Build time:         2019-09-16T07:13:07+00:00
Java Version:       11
Java VM:            Ubuntu, Version 11.0.4+11-post-Ubuntu-1ubuntu218.04.3
Operating system:   Linux, Version 4.15.18-12-pve
Environment:        amd64, 1 processors, 29203 MiB memory reserved for allocations
System components initialization in progress
10:45:42.085 [main] INFO  LINSTOR/Controller - SYSTEM - Log directory set to: '/logs'
10:45:42.175 [main] INFO  LINSTOR/Controller - SYSTEM - Linstor configuration file loaded from '/config/linstor.toml'.
10:45:42.176 [Main] INFO  LINSTOR/Controller - SYSTEM - Loading API classes started.
10:45:42.717 [Main] INFO  LINSTOR/Controller - SYSTEM - API classes loading finished: 541ms
10:45:42.717 [Main] INFO  LINSTOR/Controller - SYSTEM - Dependency injection started.
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/linstor-server/lib/guice-4.2.2.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
10:45:44.232 [Main] INFO  LINSTOR/Controller - SYSTEM - Dependency injection finished: 1515ms
10:45:44.649 [Main] INFO  LINSTOR/Controller - SYSTEM - Initializing authentication subsystem
10:45:44.875 [Main] INFO  LINSTOR/Controller - SYSTEM - Initializing the database connection pool
10:45:45.195 [Main] INFO  org.flywaydb.core.internal.license.VersionPrinter - Flyway Community Edition 5.2.4 by Boxfuse
10:45:45.212 [Main] INFO  org.flywaydb.core.internal.database.DatabaseFactory - Database: jdbc:postgresql://linstor-db-stolon-proxy/linstor2 (PostgreSQL 10.5)
10:45:45.297 [Main] INFO  org.flywaydb.core.internal.command.DbValidate - Successfully validated 26 migrations (execution time 00:00.035s)
10:45:45.317 [Main] INFO  org.flywaydb.core.internal.command.DbMigrate - Current version of schema "LINSTOR": 2019.09.09.01.01
10:45:45.317 [Main] WARN  org.flywaydb.core.internal.command.DbMigrate - outOfOrder mode is active. Migration of schema "LINSTOR" may not be reproducible.
10:45:45.318 [Main] INFO  org.flywaydb.core.internal.command.DbMigrate - Schema "LINSTOR" is up to date. No migration necessary.
10:45:45.386 [Main] INFO  LINSTOR/Controller - SYSTEM - Loading security objects
10:45:45.472 [Main] INFO  LINSTOR/Controller - SYSTEM - Current security level is NO_SECURITY
10:45:45.502 [Main] INFO  LINSTOR/Controller - SYSTEM - Core objects load from database is in progress
10:46:00.572 [Main] INFO  LINSTOR/Controller - SYSTEM - Core objects load from database completed
10:46:00.668 [Main] INFO  LINSTOR/Controller - SYSTEM - Starting service instance 'DatabaseService' of type DatabaseService
10:46:00.668 [Main] INFO  LINSTOR/Controller - SYSTEM - Starting service instance 'GrizzlyHttpServer' of type Grizzly-HTTP-Server
Oct 24, 2019 10:46:01 AM org.glassfish.grizzly.http.server.accesslog.FileAppender <init>
INFO: Access log file "/logs/rest-access.log" opened
Oct 24, 2019 10:46:02 AM org.glassfish.grizzly.http.server.NetworkListener start
INFO: Started listener bound to [127.0.0.1:3370]
Oct 24, 2019 10:46:02 AM org.glassfish.grizzly.http.server.HttpServer start
INFO: [HttpServer] Started.
10:46:02.017 [Main] INFO  LINSTOR/Controller - SYSTEM - Starting service instance 'TaskScheduleService' of type TaskScheduleService
10:46:02.017 [Main] INFO  LINSTOR/Controller - SYSTEM - Starting service instance 'TimerEventService' of type TimerEventService
10:46:02.019 [Main] INFO  LINSTOR/Controller - SYSTEM - Initializing network communications services
10:46:02.019 [Main] WARN  LINSTOR/Controller - SYSTEM - The SSL network communication service 'DebugSslConnector' could not be started because the keyStore file (/config/ssl/keystore.jks) is missing
10:46:02.020 [TaskScheduleService] INFO  LINSTOR/Controller - SYSTEM - LogArchive: Running log archive on directory: /logs
10:46:02.025 [Main] INFO  LINSTOR/Controller - SYSTEM - Created network communication service 'PlainConnector', bound to 127.0.0.1:3376
10:46:02.026 [Main] WARN  LINSTOR/Controller - SYSTEM - The SSL network communication service 'SslConnector' could not be started because the keyStore file (/config/ssl/keystore.jks) is missing
10:46:02.030 [Main] INFO  LINSTOR/Controller - SYSTEM - Reconnecting to previously known nodes
10:46:02.044 [TaskScheduleService] INFO  LINSTOR/Controller - SYSTEM - LogArchive: No logs to archive.
10:46:02.320 [Main] INFO  LINSTOR/Controller - SYSTEM - Reconnect requests sent
10:46:02.321 [Main] INFO  LINSTOR/Controller - SYSTEM - Controller initialized
10:46:08.613 [MainWorkerPool-1] WARN  LINSTOR/Controller - SYSTEM - RetryTask: Failed resource 'opennebula-data-opennebula-db-0' of node 'pve2' added for retry.
10:46:08.624 [MainWorkerPool-1] WARN  LINSTOR/Controller - SYSTEM - RetryTask: Failed resource 'opennebula-data-opennebula-db-0' of node 'pve1' added for retry.
```

**Comment by raltnoeder** [comment]
Overloading might actually be a valid problem, because it should not be possible for that to happen. The controller should just stop accepting connections and reading data at some point until some existing connections are closed, which would typically throttle the clients (because they'd be waiting for the connection to be accepted or their data to be read), but the controller certainly should not freeze.

Unfortunately, since the introduction of the REST API, our control over the robustness of the protocol is very limited, since it is all based on third-party componentes (connector, webserver, JSON libraries), and neither HTTP nor JSON were designed for robustness or scalability, which is the main reason, why I'm not a fan. Apparently, the majority of people in the Unix/Linux and Windows communities seem to have other priorities...

**Comment by ghernadi** [comment]
I am having troubles reproducing this issue. I have tested on our internal most upToDate master branch, but also on your `1.1.2` release version (btw, `1.2.0` is already available).
Please review my test-case and tell me if you see something which could be different to what you are doing


```
linstor n c bravo
linstor n c charlie
linstor n c delta
linstor n l
linstor sp c lvm bravo lvmpool scratch
linstor sp c lvm charlie lvmpool scratch
linstor sp c lvm delta lvmpool scratch
linstor rd c rsc
linstor vd c rsc 100m
linstor r c bravo rsc -s lvmpool
linstor r c charlie rsc -s lvmpool
linstor r c delta rsc -s lvmpool
sleep 1.0s
linstor r lv
ssh root@bravo mkfs.ext4 -Enodiscard /dev/drbd1000
ssh root@bravo mkdir -p /mnt/ext
ssh root@bravo mount /dev/drbd1000 /mnt/ext
linstor v l
+-----------------------------------------------------------------------------------------------------------------+
| Node    | Resource | StoragePool | VolumeNr | MinorNr | DeviceName    | Allocated | InUse  |              State |
|=================================================================================================================|
| bravo   | rsc      | lvmpool     | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [1;32mInUse [0m | [0;32m          UpToDate[0m |
| charlie | rsc      | lvmpool     | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [0;32mUnused[0m | [0;32m          UpToDate[0m |
| delta   | rsc      | lvmpool     | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [0;32mUnused[0m | [1;33mSyncTarget(50.59%)[0m |
+-----------------------------------------------------------------------------------------------------------------+

linstor r td bravo rsc -d
linstor v l
+----------------------------------------------------------------------------------------------------------------+
| Node    | Resource | StoragePool          | VolumeNr | MinorNr | DeviceName    | Allocated | InUse  |    State |
|================================================================================================================|
| bravo   | rsc      | DfltDisklessStorPool | 0        | 1000    | /dev/drbd1000 |           | [1;32mInUse [0m | [0;32mDiskless[0m |
| charlie | rsc      | lvmpool              | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [0;32mUnused[0m | [0;32mUpToDate[0m |
| delta   | rsc      | lvmpool              | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [0;32mUnused[0m | [0;32mUpToDate[0m |
+----------------------------------------------------------------------------------------------------------------+

linstor r td bravo rsc -s lvmpool
linstor v l
+-----------------------------------------------------------------------------------------------------------+
| Node    | Resource | StoragePool | VolumeNr | MinorNr | DeviceName    | Allocated | InUse  |        State |
|===========================================================================================================|
| bravo   | rsc      | lvmpool     | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [1;32mInUse [0m | [1;31mInconsistent[0m |
| charlie | rsc      | lvmpool     | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [0;32mUnused[0m | [0;32m    UpToDate[0m |
| delta   | rsc      | lvmpool     | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [0;32mUnused[0m | [0;32m    UpToDate[0m |
+-----------------------------------------------------------------------------------------------------------+

sleep 2.0s
linstor v l
+-------------------------------------------------------------------------------------------------------+
| Node    | Resource | StoragePool | VolumeNr | MinorNr | DeviceName    | Allocated | InUse  |    State |
|=======================================================================================================|
| bravo   | rsc      | lvmpool     | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [1;32mInUse [0m | [0;32mUpToDate[0m |
| charlie | rsc      | lvmpool     | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [0;32mUnused[0m | [0;32mUpToDate[0m |
| delta   | rsc      | lvmpool     | 0        | 1000    | /dev/drbd1000 | 104 MiB   | [0;32mUnused[0m | [0;32mUpToDate[0m |
+-------------------------------------------------------------------------------------------------------+

ssh root@bravo umount /mnt/ext
```

**Comment by kvaps** [comment]
Agree, your case is working,

I able to reproduce if lvmpool have `PrefNic` property set but `DfltDisklessStorPool` have no

**Comment by ghernadi** [comment]
You are right. I am still unsure why or how, but apparently the `PrefNic` indeed is causing an issue. I will look into this (even with our internal master, i.e. `1.2.0+`)

**Comment by ghernadi** [comment]
_Note to myself: thinking helps._
If you think about it, it makes sense. 

When you are executing a `toggle-disk`, the following things are changing:
* DRBD goes from diskful to diskless
* the storagepool changes from `data` to `DfltDisklessStorPool`
  * this also changes the IP (unless the same `PrefNic` is set, which is not as you said). 

The next `drbdadm adjust` will now tear down the DRBD-connection to the other DRBD-resources (because the IP has changed), and try to make its own local DRBD resource diskless. These two things do not go well with each other, as the error message `Need access to UpToDate data` confirms.

However, fixing this sounds quite troublesome. We would have to split up the `toggle-disk` command internally into two steps, first waiting for the DRBD-resource to establish connection over the new IP and after that execute the actual toggle disk.  
Although we should have the option to do that on either the controller side OR the satellite side (latter could be a bit more "atomic" regarding our internal async-communication), but I guess I am going a bit too deep into implementation details right now.

So to summarize: You are right. This is a bug. (Now) this is also an understood bug. But currently we have quite a few tasks with higher priority, so I can't tell you for sure when we will have time to fix this. If you are still interested in solving this issue in a timely fashion, feel free to contact Linbit directly with this feature request.

Best regards,
Gabor

**Comment by kvaps** [comment]
Hey, no problem, I'm glad that we found root of this problem, in my case it was just misconfiguration for DfltDisklessStorPool pool and I wasn't able to recognize due https://github.com/LINBIT/linstor-server/issues/42

BTW we already another diskless pool with different name and all parameters set.

So we just need to check connection before toggle disk for showing reliable error to user as you mentioned.

Thanks!
